Fixed #32042 -- Improved error messages for the number of submitted forms in formsets.

meghanabhange opened the pull request:
PR to fix the messaging issue mentioned in ticket #32042 - https://code.djangoproject.com/ticket/32042

felixxm reviewed (commented):
@meghanabhange Thanks for this patch :+1:  I would change the current messages instead of introducing a new one, e.g.

- `too_many_forms`: _"Please submit at most %d forms."_
- `too_few_forms`: _"Please submit at least %d forms."_

this would be consistent with the current behavior of other min/max validators.

meghanabhange commented:
Yes @felixxm, made those changes in 21f683f

- Changed the error messages. 
- Changed it even in tests where the older message was being referenced. 

Does this solution look good?

felixxm commented on django/forms/formsets.py:
```diff
@@ -342,15 +342,15 @@ def full_clean(self):
                 raise ValidationError(ngettext(
-                    "Please submit %d or fewer forms.",
-                    "Please submit %d or fewer forms.", self.max_num) % self.max_num,
+                    "Please submit at most %d forms.",
```
The singular form should be `Please submit at most %d form.`.

claudep replied:
I'm no English specialist, but shouldn't we remove the `s` in the first form, because reading `Please submit at most 1 forms.` sounds weird with the final `s`.

felixxm commented on django/forms/formsets.py:
```diff
@@ -342,15 +342,15 @@ def full_clean(self):
                 raise ValidationError(ngettext(
-                    "Please submit %d or more forms.",
-                    "Please submit %d or more forms.", self.min_num) % self.min_num,
+                    "Please submit at least %d forms.",
```
The singular form should be `Please submit at least %d form.`.

felixxm commented on tests/model_formsets/tests.py:
```diff
@@ -1267,14 +1267,66 @@ def test_modelformset_validate_max_flag(self):
+    def test_modelformset_min_num_equals_max_num_less_than(self):
+        # If validate_max and validate_min is set and max_num is less than
+        # TOTAL_FORMS in the data, then this throws an exception.
+
```
Chop blank line.

felixxm commented on tests/model_formsets/tests.py:
```diff
@@ -1267,14 +1267,66 @@ def test_modelformset_validate_max_flag(self):
+        data = {
+            'form-TOTAL_FORMS': '3',
+            'form-INITIAL_FORMS': '0',
+            'form-MAX_NUM_FORMS': '2',  # should be ignored
```
We can chop this comment, it's not related with tested behavior (the same in the second test).

felixxm commented on tests/model_formsets/tests.py:
```diff
@@ -1267,14 +1267,66 @@ def test_modelformset_validate_max_flag(self):
+            'form-1-price': '24.00',
+            'form-1-quantity': '2',
+            'form-2-price': '36.00',
+            'form-2-quantity': '3'
```
Add trailing comma.

felixxm commented on tests/model_formsets/tests.py:
```diff
@@ -1267,14 +1267,66 @@ def test_modelformset_validate_max_flag(self):
+            max_num=2,
+            validate_max=True,
+            min_num=2,
+            validate_min=True)
```
Use hanging indentation (the same in the second test).

felixxm reviewed (commented):
@meghanabhange Thanks for the update :+1: I left small comments.

meghanabhange commented:
The singular form of the message completely slipped my mind. Fixed that and the formatting changes in c5245d9

felixxm commented:
@meghanabhange Thanks for updates :+1: Welcome aboard :boat: 

I simplified new tests by using a single field model.
